src/main/java/com/google/cloud/spanner/pgadapter/statements/AbstractFetchOrMoveStatement.java

```diff
@@ -0,0 +1,269 @@
+    if (!this.executed) {
+      try {
+        checkSupported();
+        int count;
+        if (fetchOrMoveStatement.count == null) {
+          count =
+              fetchOrMoveStatement.direction == null
+                  ? 1
+                  : fetchOrMoveStatement.direction.getDefaultCount();
+        } else {
+          count = fetchOrMoveStatement.count;
+        }
+        execute(backendConnection, count);
+        // Set a null result to indicate that this statement should not return any result.
+        setFutureStatementResult(Futures.immediateFuture(null));
+      } catch (Exception exception) {
+        setFutureStatementResult(Futures.immediateFailedFuture(exception));
+      }
+    }
+  }
+
+  protected void checkSupported() throws PGException {
+    if (fetchOrMoveStatement.count != null && fetchOrMoveStatement.count <= 0) {
+      throw PGExceptionFactory.newPGException(
+          "only positive count is supported", SQLState.FeatureNotSupported);
+    }
+    if (fetchOrMoveStatement.direction != null && !fetchOrMoveStatement.direction.isSupported()) {
+      throw PGExceptionFactory.newPGException(
+          fetchOrMoveStatement.direction + " is not supported", SQLState.FeatureNotSupported);
+    }
+  }
+
+  protected abstract void execute(BackendConnection backendConnection, int count) throws Exception;
+
+  @Override
+  public StatementType getStatementType() {
+    return StatementType.CLIENT_SIDE;
+  }
+
+  @Override
+  public Future<StatementResult> describeAsync(BackendConnection backendConnection) {
+    // Return null to indicate that the FETCH or MOVE statement does not return any
+    // RowDescriptionResponse.
+    return Futures.immediateFuture(null);
+  }
+
+  @Override
+  public IntermediatePortalStatement createPortal(
+      String name,
+      byte[][] parameters,
+      List<Short> parameterFormatCodes,
+      List<Short> resultFormatCodes) {
+    // FETCH and MOVE do not support binding any parameters, so we just return the same statement.
+    return this;
+  }
+
+  static <T extends ParsedFetchOrMoveStatement> T parse(String sql, String type, Class<T> clazz) {
+    Preconditions.checkNotNull(sql);
+    Preconditions.checkNotNull(type);
+
+    // {MOVE | FETCH} [ direction ] [ FROM | IN ] cursor_name
+    SimpleParser parser = new SimpleParser(sql);
+    if (!parser.eatKeyword(type)) {
+      throw PGExceptionFactory.newPGException(
+          "not a valid " + type.toUpperCase() + " statement: " + sql, SQLState.SyntaxError);
+    }
+    Direction direction;
+    if (parser.eatKeyword("forward")) {
+      if (parser.eatKeyword("all")) {
+        direction = Direction.FORWARD_ALL;
+      } else {
+        direction = Direction.FORWARD;
+      }
+    } else if (parser.eatKeyword("backward")) {
+      if (parser.eatKeyword("all")) {
+        direction = Direction.BACKWARD_ALL;
+      } else {
+        direction = Direction.BACKWARD;
+      }
+    } else {
```

Comment:
Do need to add a validation for ill formed query here?

Reply:
No, that handles itself automatically. I've added the following comment to clarify:

```
              // This ensures that the stream will return the first valid direction keyword that it
              // finds, or null if no valid direction keyword is found. If no valid direction
              // keyword is found, then the position of the parser will also not be moved.
```

So basically: If no valid direction keyword is found, then direction is null (which could be valid, as direction is optional). The further parsing of the statement might fail, as it expects for example a count or one of the `from | in` keywords to follow.